File: python/algorithms/math/fast_power/fast_power.py
"""快速幂：二进制幂；模运算版为刷题常用入口。"""

from __future__ import annotations


def fast_pow(a: int, e: int) -> int:
    """a^e（非模；指数很大时注意溢出）。"""
    r = 1
    base = a
    while e:
        if e & 1:
            r *= base
        base *= base
        e >>= 1
    return r


def mod_pow(a: int, e: int, mod: int) -> int:
    """a^e mod mod。"""
    if mod == 1:
        return 0
    r = 1 % mod
    a %= mod
    while e:
        if e & 1:
            r = (r * a) % mod
        a = (a * a) % mod
        e >>= 1
    return r


if __name__ == "__main__":
    assert fast_pow(2, 10) == 1024
    assert fast_pow(3, 0) == 1
    assert mod_pow(7, 10, 13) == 4
    assert mod_pow(2, 100, 1_000_000_007) == 976371285
    assert mod_pow(7, 0, 1) == 0
    print("fast_power OK")

File: cpp/algorithms/math/fast_power/fast_power.cpp
// 快速幂：二进制幂与模幂
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

long long fastPow(long long a, long long e) {
    long long r = 1;
    while (e) {
        if (e & 1) r *= a;
        a *= a;
        e >>= 1;
    }
    return r;
}

long long modPow(long long a, long long e, long long mod) {
    if (mod == 1) return 0;
    long long r = 1 % mod;
    a %= mod;
    while (e) {
        if (e & 1) r = (r * a) % mod;
        a = (a * a) % mod;
        e >>= 1;
    }
    return r;
}

int main() {
    assert(fastPow(2, 10) == 1024);
    assert(fastPow(3, 0) == 1);
    assert(modPow(7, 10, 13) == 4);
    assert(modPow(2, 100, 1'000'000'007LL) == 976371285LL);
    cout << "fast_power OK" << endl;
    return 0;
}
